server/lib/dataone/dataone_provider.py

```diff
@@ -0,0 +1,115 @@
+from girder import logger
+
+from ..import_providers import ImportProvider
+from ..data_map import DataMap
+from ..file_map import FileMap
+from ..import_item import ImportItem
+from ..entity import Entity
+from .dataone_register import D1_lookup, get_package_list
+from .dataone_register import \
+    extract_metadata_docs, \
+    get_documents, \
+    extract_data_docs, \
+    extract_resource_docs, \
+    check_multiple_metadata
+from ...constants import DataONELocations
+
+ALL_LOCATIONS = [DataONELocations.prod_cn, DataONELocations.dev_mn, DataONELocations.dev_cn]
+ALL_LOCATIONS_2 = ['https://knb.ecoinformatics.org/#view/', 'https://search.dataone.org/view/',
+                   'https://cn.dataone.org/cn/v2/resolve/']
+
+
+class DataOneImportProvider(ImportProvider):
+    def __init__(self):
+        super().__init__('DataONE')
+
+    def matches(self, entity: Entity) -> bool:
+        # need to deal with knb.ecoinformatics.org and other such places since, e.g.,
+        # doi:10.5063/F1JM27VG points there
+        url = entity.getValue()
+        for base_url in ALL_LOCATIONS_2:
+            if url.startswith(base_url):
+                return True
```

Comment:
I don't think we should assume that only DataONE objects are associated with uuids and it in a URN. This probably goes back to the idea that for some type of identifiers, we need to do a lookup through the actual provider.

Reply:
That also made me realize that we break our flagship example [here](https://wholetale.readthedocs.io/users_guide/quickstart.html#composing-a-new-tale). Trying to register `10.5065/D6862DM8` would probably end up in null provider...

Reply:
Well, I didn't recognize that https://arcticdata.io/catalog/#view/doi:10.5065/D6862DM8 is somehow connected to DataONE. Is this dataset also in one of the canonical DataONE sites?

Reply:
I don't know what they're considering as canonical repos, but this dataset can be found via https://search.dataone.org/view/doi:10.5065/D6862DM8

Reply:
This kinda came up on our call today but yeah: The DOI `10.5065/D6862DM8` resolves to https://arcticdata.io/catalog/#view/doi:10.5065/D6862DM8 and arcticdata.io is a member node in DataONE. https://arcticdata.io/catalog/#view/doi:10.5065/D6862DM8 is an HTML representation of a dataset with the DataONE identifier `doi:10.5065/D6862DM8` (which just happens to look like a DOI but is actually just an internal identifier that looks like a DOI) which corresponds to a file (DataONE Object) that is replicated within DataONE in two places:

```sh
❯ curl "https://cn.dataone.org/cn/v2/resolve/doi:10.5065/D6862DM8"
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<ns2:objectLocationList xmlns:ns2="http://ns.dataone.org/service/types/v1">
    <identifier>doi:10.5065/D6862DM8</identifier>
    <objectLocation>
        <nodeIdentifier>urn:node:CN</nodeIdentifier>
        <baseURL>https://cn.dataone.org/cn</baseURL>
        <version>v1</version>
        <version>v2</version>
        <url>https://cn.dataone.org/cn/v2/object/doi:10.5065%2FD6862DM8</url>
    </objectLocation>
    <objectLocation>
        <nodeIdentifier>urn:node:ARCTIC</nodeIdentifier>
        <baseURL>https://arcticdata.io/metacat/d1/mn</baseURL>
        <version>v1</version>
        <version>v2</version>
        <url>https://arcticdata.io/metacat/d1/mn/v2/object/doi:10.5065%2FD6862DM8</url>
    </objectLocation>
</ns2:objectLocationList>
```

Reply:
"They copy and paste the identifier on the page urn:uuid:30a0b05a-b73f-4ec5-bf2c-285c6254b1a9"

That could very well be an identifier for some other provider. When I look at it, it tells me nothing. If I google for it (actually you should try, it's funny) I get 5 results about planes and a mention that none of my search terms were found. What is google searching for exactly remains a mystery.

If we go that route in general, our identification step (answering the question "what provider does this belong to?"), it would be, on average, an O(n) query in the number of providers n. I don't like that. I think O(1) is something that scales, O(n) is not. That said, we could try the O(1) search first and fall back to querying providers one by one.

I think what's reasonable and covers most cases is:
  * DOI from a paper
  * URL of a page that represents a dataset
  * maybe "dataone:<D1Identifier>"

Reply:
Avoiding O(n) stuff is 👍 w/ me.

Re:

> - DOI from a paper
> - URL of a page that represents a dataset
> - maybe "dataone:"

The first two sound fine. The third one would never happen since we don't expose identifiers under that URI scheme. That said, the first two still leave _what_ to do open.

Could we avoid some of these issues by always making the user specify which registration method they intended? Might simplify things.

Reply:
> we can automatically tell what provider that URL belongs to

Oh, okay. I thought this was a problem after reading the above conversation. So is this all sorted?